lib/date-parse/index.js

```diff
@@ -0,0 +1,169 @@
+const AWS = require('aws-sdk')
+const SierraItem = require('pcdm-store-updater/lib/models/item-sierra-record')
+
+const { preparsingObjects, returnOneYear } = require('./preparsingObjects')
+
+AWS.config.region = 'us-east-1'
+
+let lambda
+
+function lambdaClient () {
+  if (!lambda) lambda = new AWS.Lambda()
+  return lambda
+}
+
+/**
+ * Given an array of bibs that already have items attached, returns an
+ * array of the fieldtagvs from all items on the bib.
+ */
+const extractFieldtagvs = (bibs) => {
+  return bibs
+    .map((bib) => {
+      return bib.items.map((item) => {
+        return (new SierraItem(item)).fieldTag('v')[0]
+      })
+    })
+    .flat()
+    .filter((fieldtagv) => fieldtagv)
+}
+
+/**
+ *
+ * @param {*} batchSize a number
+ * @param {*} originalArray array
+ * @returns a nested array whose elements have a maximum
+ *  length of batchsize
+ */
+
+const batchDates = (batchSize, originalArray) => {
+  const batches = []
+  for (let i = 0; i < originalArray.length; i += batchSize) {
+    batches.push(originalArray.slice(i, i + batchSize))
+  }
+  return batches
+}
+
+let dateCache = {}
+
+const checkCache = (fieldtagv) => {
+  return dateCache[fieldtagv]
+}
+
+/**
+ *
+ * @param {*} dates, an array of strings representing field tag 'v'
+ * @returns an array of those dates with some string manipulation
+ *  so the data plays nicely with timetwister
+ */
+
+const preparse = function (dates) {
+  return dates.map((date) => {
+    // do some string manipulation so data works better with timetwister lambda
+    preparsingObjects.forEach((preparse) => {
+      if (date.match(preparse.matchExpression)) {
+        date = preparse.transform(date)
+      }
+    })
+    return date
+  })
+}
+
+/**
+ *
+ * @param {array} bibs an array of bibs with items attached
+ * @returns nothing, used for side effect of calling timetwister lambda and filling cache
+ */
+
+const parseDatesAndCache = async function (bibs) {
+  const dates = extractFieldtagvs(bibs)
+  const preparsedDates = preparse(dates)
+  try {
+    let ranges = await timeTwist(preparsedDates)
+    ranges = filterNulls(ranges, preparsedDates)
+    dateCache = dates.reduce((cache, date, i) => {
+      return { ...cache, [date]: ranges[i] }
+    }, {})
```

Comment:
Can we add time-ellapsed reporting to this call? I predict we'll want to be able to inspect slow calls. Something like:
```
logger.debug(`parseDatesAndCache: Took {time}ms to process ${dates.length} from ${bibs.length} bibs`)
```

Reply:
Lines 79 - 82 are basically exactly what `_parseDates` does. Couldn't you define `parseDatesAndCache` like this?
```
const parseDatesAndCache = async function (bibs) {
  const dates = extractFieldtagvs(bibs)
  const ranges = _parseDates(dates)
  
  // Update cache:
  dateCache = dates.reduce((cache, date, i) => {
    return { ...cache, [date]: ranges[i] }
  }, {})
}
```

Reply:
updated!